view/frontend/web/internals/common.js

```diff
@@ -1,3 +1,47 @@
+var algolia = {
+	allowedHooks: [
+		'beforeAutocompleteSources',
+		'beforeAutocompleteOptions',
+		'beforeInstantsearchInit',
+		'beforeWidgetInitialization',
+		'beforeInstantsearchStart',
+		'afterInstantsearchStart'
+	],
+	registeredHooks: [],
+	registerHook: function (hookName, callback) {
+		if (this.allowedHooks.indexOf(hookName) === -1) {
+			throw 'Hook "' + hookName + '" cannot be defined. Please use one of ' + this.allowedHooks.join(', ');
+		}
+		
+		if (!this.registeredHooks[hookName]) {
+			this.registeredHooks[hookName] = [callback];
+		} else {
+			this.registeredHooks[hookName].push(callback);
+		}
+	},
+	getRegisteredHooks: function(hookName) {
+		if (this.allowedHooks.indexOf(hookName) === -1) {
+			throw 'Hook "' + hookName + '" cannot be defined. Please use one of ' + this.allowedHooks.join(', ');
+		}
+		
+		if (!this.registeredHooks[hookName]) {
+			return [];
+		}
+		
+		return this.registeredHooks[hookName];
+	},
+	triggerHooks: function () {
+		var hookName = arguments[0],
+			clearedArguments = Array.prototype.slice.call(arguments, 1);
+		
+		this.getRegisteredHooks(hookName).forEach(function(callback) {
+			clearedArguments[0] = callback.apply(this, clearedArguments);
```

Comment:
can you explain the usage of clearedArguments here? What's the contract of the hook?